src/main/java/com/proofpoint/discovery/IpHostnameAuthManager.java

```diff
@@ -0,0 +1,189 @@
+        }
+        executor.shutdownNow();
+        future = null;
+    }
+
+    private synchronized void startDiscoveryRefresh()
+    {
+        if (future == null) {
+            future = executor.scheduleWithFixedDelay(() -> {
+                try {
+                    updateDiscoveryAddrs(discoverySelector.selectAllServices());
+                }
+                catch (Throwable e) {
+                    logger.warn("Unable to refresh discovery servers. No replication will be permitted");
+                }
+            }, 0, updateInterval.toMillis(), TimeUnit.MILLISECONDS);
+        }
+    }
+
+    private void updateDiscoveryAddrs(List<ServiceDescriptor> descriptors)
+    {
+        Set<InetAddress> otherDiscoveryHosts = descriptors.stream()
+                .filter(d -> !localNodeId.equals(d.getNodeId()))
+                .map(d -> d.getProperties().get("http"))
+                .map(URI::create)
+                .map(u -> {
+                    try {
+                        return InetAddress.getAllByName(u.getHost());
+                    }
+                    catch (UnknownHostException e) {
+                        return null;
+                    }
+                })
+                .filter(Objects::nonNull)
+                .flatMap(Arrays::stream)
+                .collect(Collectors.toSet());
+        discoveryAddrs.removeIf(entry -> !otherDiscoveryHosts.contains(entry));
+        discoveryAddrs.addAll(otherDiscoveryHosts);
+    }
+
+    @Override
+    public void checkAuthAnnounce(Id<Node> nodeId, DynamicAnnouncement announcement, HttpServletRequest request)
+    {
+        try {
+            InetAddress requesterAddress = InetAddress.getByName(request.getRemoteAddr());
+            String nodeAnnouncer = dynamicStore.getAnnouncer(nodeId);
+            if (mismatchedAnnouncer(nodeAnnouncer, requesterAddress)) {
+                //NodeId was previously announced by a different IP
+                logger.warn("IP %s tried to re-announce node %s owned by IP %s", requesterAddress.getHostAddress(), nodeId.toString(), nodeAnnouncer);
+                throw new ForbiddenException();
+            }
+            Set<DynamicServiceAnnouncement> newAnnouncements = announcement.getServiceAnnouncements();
+            for (DynamicServiceAnnouncement newAnnouncement : newAnnouncements) {
+                Map<String, String> properties = newAnnouncement.getProperties();
+                if (invalidHostnameIp(properties.get("http"), requesterAddress) ||
+                        invalidHostnameIp(properties.get("https"), requesterAddress) ||
+                        invalidHostnameIp(properties.get("admin"), requesterAddress)) {
+                    //Service is announcing a host with a different set of IPs
+                    throw new ForbiddenException();
+                }
+            }
+        }
+        catch (UnknownHostException e) {
+            //Unable to validate an IP or look up a hostname
+            logger.error(e);
```

Comment:
We should not be logging at error level in response to client input.

These would be more appropriately logged to the audit log.